### PR TITLE
IGNITE-23665 Sql. Introduce query execution program

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-23665

----------
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/ExecutionPhase.java [hunk 1]
```diff
@@ -65,9 +47,4 @@ public enum ExecutionPhase {
     Result evaluate(Query query) {
         return handler.handle(query);
     }
-
-    /** Returns {@code true} if transition is allowed from current phase to given one, return {@code false} otherwise. */
-    boolean transitionAllowed(ExecutionPhase newPhase) {
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/ExecutionPhase.java:70]: ok, got it, seems this check is redundant for now

[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/ExecutionPhase.java:70]: possible transitions now defined by program (which is initialized with collection of `Transition`'s object). The check is moved to Program as well (see `org.apache.ignite.internal.sql.engine.exec.fsm.Program#advanceQuery`). So, it's not up to ExecutionPhase whether transition is allowed or not.

**Review by AMashenkov — APPROVED** [review]

**Review by zstan — APPROVED** [review]
